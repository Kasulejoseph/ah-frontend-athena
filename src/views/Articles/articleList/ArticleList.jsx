import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { MDBContainer, MDBCol, MDBRow } from "mdbreact";
import ArticleItem from "../../../components/articles/articleItem/ArticleItem";
import { getArticles } from "../../../actions/articleActions/ArticleActions";
import getTags from "../../../actions/tagsAction";
import TagsForm from "../../../components/Tags/TagsForm/TagsForm";
import Paginations from "../../Pagination/Paginations";
import Search from "../../Search/Search";
import Load from "../../../components/Load/Load";
import "./ArticleList.scss";

export class ArticleList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      articles: []
    };
  }

  componentWillMount() {
    const { getArticles, getTags } = this.props;
    getTags();
    getArticles();
  }

  componentWillReceiveProps(nextProps) {
    const { articles } = nextProps;
    if (articles) {
      const results = articles;
      this.setState({ articles: results });
    }
  }

  updateArticleList = newArticles => {
    this.setState({ articles: newArticles });
  };

  render() {
    const { articles, totalPages } = this.state;
    const { fullArticle, tags } = this.props;
    let { pageNumber } = this.state;
    pageNumber = [];
    for (let i = 1; i <= totalPages; i++) {
      pageNumber.push(i);
    }

    if (articles.length === 0) {
      return (
        <div>
          <h1
            className="d-flex justify-content-center"
            style={{ marginTop: "20%" }}
          >
            There are no articles for this search
          </h1>
        </div>
      );
    } else if (articles) {
      const art = articles.map(article => {
        return <ArticleItem key={article.id} article={article} />;
      });
      return (
        <MDBContainer className="pl-3">
          <MDBRow>
            <MDBCol size="9" className="w-100 m-0">
              {art}
              <Paginations
                fullArticle={fullArticle}
                update={this.updateArticleList}
              />
            </MDBCol>
            <MDBCol size="3" className="position-sticky m1">
              <Search />
              <TagsForm tags={tags} />
            </MDBCol>
          </MDBRow>
        </MDBContainer>
      );
    } else {
      return <Load />;
    }
  }
}

const mapStateToProps = state => {
  return {
    articles: state.articles.articles,
    tags: state.tags.tags,
    moreArticles: state.articles.articles,
    currentPage: state.articles.currentPage,
    totalPages: state.articles.totalPages,
    fullArticle: state.articles.fullArticle
  };
};

ArticleList.propTypes = {
  getArticles: PropTypes.func,
  getTags: PropTypes.func,
  results: PropTypes.shape({}),
  tags: PropTypes.shape([]),
  articles: PropTypes.shape([]),
  fullArticle: PropTypes.shape({})
};

ArticleList.defaultProps = {
  getArticles: () => {},
  getTags: () => {},
  articles: [],
  results: {},
  tags: [],
  fullArticle: {}
};

export default connect(
  mapStateToProps,
  { getArticles, getTags }
)(ArticleList);