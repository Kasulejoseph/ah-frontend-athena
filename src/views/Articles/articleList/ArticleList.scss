#tags-position {
  position: absolute !important;
  width: 40vw !important;
  top: 0 !important;
  right: 10px !important;
}

.center-parent {
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
}
