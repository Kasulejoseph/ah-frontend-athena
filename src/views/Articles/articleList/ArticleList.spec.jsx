import React from "react";
import { shallow } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { ArticleList } from "./ArticleList";

jest.mock("react-router-dom");
let newArticle = {
  title: "title",
  description: "description",
  body: "this is body",
  author: {
    username: "henry"
  }
};

let articleDta = {
  articles: {
    results: [
      {
        title: "title",
        description: "description",
        body: "this is body",
        author: {
          username: "henry"
        }
      }
    ]
  }
};

const intitialState = {
  articles: []
};

describe("<ArticleList>", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<ArticleList {...newArticle} />);
  });
  it("matches snapshort", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should mount article Item", () => {
    const mockStore = configureStore();
    let store = mockStore(intitialState);
    let wrapp = shallow(
      <Provider store={store}>
        <ArticleList {...articleDta} />
      </Provider>
    );
    expect(wrapp.state("storeState")).toEqual({articles: []})
  });

  it("should handle empty articles in next props properly", () => {
    wrapper.setProps({ articles: [] });
    expect(wrapper.instance().props.articles).toEqual([]);
  });
});

