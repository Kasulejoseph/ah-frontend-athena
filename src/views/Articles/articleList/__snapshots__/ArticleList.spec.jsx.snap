// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleList> matches snapshort 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ArticleList
    articles={Array []}
    author={
      Object {
        "username": "henry",
      }
    }
    body="this is body"
    description="description"
    fullArticle={Object {}}
    getArticles={[Function]}
    getTags={[Function]}
    results={Object {}}
    tags={Array []}
    title="title"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h1
        className="d-flex justify-content-center"
        style={
          Object {
            "marginTop": "20%",
          }
        }
      >
        There are no articles for this search
      </h1>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "There are no articles for this search",
        "className": "d-flex justify-content-center",
        "style": Object {
          "marginTop": "20%",
        },
      },
      "ref": null,
      "rendered": "There are no articles for this search",
      "type": "h1",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h1
          className="d-flex justify-content-center"
          style={
            Object {
              "marginTop": "20%",
            }
          }
        >
          There are no articles for this search
        </h1>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "There are no articles for this search",
          "className": "d-flex justify-content-center",
          "style": Object {
            "marginTop": "20%",
          },
        },
        "ref": null,
        "rendered": "There are no articles for this search",
        "type": "h1",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
